keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,390 @@
+        completions: 1 # number of nodes per pool
+        backoffLimit: 0
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+              
+```
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Proposed Failure Policy API:
+
+```go
+
+// FailurePolicyAction defines the action the JobSet controller will take for
+// a given FailurePolicyRule.
+type FailurePolicyAction string
+
+const (
+  // Fail JobSet immediately, regardless of maxRestarts.
+  FailJobSet FailurePolicyAction = "FailJobSet"
+
+  // Don't count the failure against maxRestarts.
+  Ignore FailurePolicyAction = "Ignore"
+
+  // The failed child job is permanently failed and should not be
+  // restarted, active child jobs will continue to run to completion.
+  FailJob FailurePolicyAction = "FailJob"
+
+  // Restart the failed child job only, not the whole JobSet.
+  RestartJob FailurePolicyAction = "RestartJob"
+)
+
+// FailurePolicyRule defines a FailurePolicyAction to be executed if a child job
+// fails due to a reason listed in OnJobFailureReasons.
+// Only only rule can match a unique (Job failure reason, ReplicatedJob) pair.
+type FailurePolicyRule struct {
+  // The action to take if the rule is matched.
+  // +kubebuilder:validation:Enum:=FailJobSet;Ignore;FailJob;RestartJob
+  Action FailurePolicyAction `json:"action"`
+  // The requirement on the job failure reasons. The requirement
+  // is satisfied if at least one reason matches the list.
+  // Each unique failure reason can only be associated with one FailurePolicyRule.
+  // An empty list applies the rule to any job failure reason.
+  // +kubebuilder:validation:UniqueItems:true
+  OnJobFailureReasons []string `json:"onJobFailureReasons"`
+  // TargetReplicatedJobs are the names of the replicated jobs the operator will apply to.
+  // A null or empty list will apply to all replicatedJobs.
+  // +optional
+  // +listType=atomic
+  // +kubebuilder:validation:UniqueItems
+  TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
+}
+
+type FailurePolicy struct {
+  // MaxRestarts defines the limit on the number of JobSet restarts.
+  // A restart is achieved by recreating all active child jobs.
+  // +optional
```

Comment:
This would actually be a breaking change. MaxRestarts was not marked as optional before. I also think since its a int32 the nil is just 0 which is probably fine.

Maybe just rules should be optional and default means that we restart with ignore?

Reply:
> This would actually be a breaking change. MaxRestarts was not marked as optional before. I also think since its a int32 the nil is just 0 which is probably fine.

Ah good catch, thanks, fixed.

> Maybe just rules should be optional and default means that we restart with ignore?

In this proposal, if no rules are specified, we fall back to the default behavior of restarting the JobSet and **not** ignoring the restart (i.e., incrementing `jobset.sigs.k8s.io/restart-attempt`) so it counts against MaxRestarts.

If the user doesn't want a restart to count against MaxRestarts, they'll need to define a FailurePolicyRule with action type Ignore, configured for any (or all) failure reasons, and any (or all) target replicated jobs.

Reply:
That second part is worth a comment in the API.